T3680: protocols: add dhclient hooks for dhcp-interface static routes

MattKobayashi opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
This PR adds `dhclient` hooks to the image that re-configure static routes configured with `dhcp-interface` when an DHCP state change occurs on one of those interfaces. The code is inspired by (and largely identical to) the existing `dhclient` hooks for IPsec interfaces. This resolves the issue reported in T3680 of static routes with `dhcp-interface` as the next hop not being added to or removed from FRRouting when DHCP is renewed on an interface. The other issues mentioned in T3680 persist.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T3680

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

```
 INFO - Executing VyOS smoketests
DEBUG - vyos@vyos:~$ /usr/bin/vyos-smoketest
DEBUG - /usr/bin/vyos-smoketest
DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py
DEBUG - test_01_static (__main__.TestProtocolsStatic.test_01_static) ... ok
DEBUG - test_02_static_table (__main__.TestProtocolsStatic.test_02_static_table) ... ok
DEBUG - test_03_static_vrf (__main__.TestProtocolsStatic.test_03_static_vrf) ... ok
DEBUG - test_04_static_multicast (__main__.TestProtocolsStatic.test_04_static_multicast) ... ok
DEBUG - test_05_dhcp_default_route (__main__.TestProtocolsStatic.test_05_dhcp_default_route) ... ok
DEBUG - test_06_dhcp_default_route_for_vrf (__main__.TestProtocolsStatic.test_06_dhcp_default_route_for_vrf) ... ok
DEBUG - test_07_dhcp_interface_static_routes (__main__.TestProtocolsStatic.test_07_dhcp_interface_static_routes) ... ok
DEBUG -
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 7 tests in 183.038s
DEBUG -
DEBUG - OK
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

hedrok commented:
Hi Matthew, thank you very much for your contribution!
VyOS team currently works on fix that will solve all issues in https://vyos.dev/T3680 and https://vyos.dev/T5811
When it's ready, we'll let you know by comment here.

c-po commented on src/conf_mode/protocols_static.py:
```diff
@@ -92,6 +94,23 @@ def generate(config_dict):
+
+    # Write the interface list for DHCP hooks or clean up if empty
+    if dhcp_interfaces:
+        with open(DHCP_HOOK_IFLIST, 'w') as f:
```
Please use `vyos.utils.write_file() instead.

MattKobayashi replied:
Resolved in https://github.com/vyos/vyos-1x/pull/4622/commits/bc715d734a59dbcf7d90761bb386dbe76f22e4e7

sever-sever commented on src/etc/dhcp/dhclient-exit-hooks.d/98-vyos-static-routes-dhclient-hook:
```diff
@@ -0,0 +1,48 @@
+fi
+
+# Best effort wait for any active commit to finish
+sudo python3 - <<PYEND
+from vyos.utils.commit import wait_for_commit_lock
+
+if __name__ == '__main__':
+    wait_for_commit_lock()
+    exit(0)
+PYEND
+
+# Now re-generate the static routes config
+sudo /usr/libexec/vyos/conf_mode/protocols_static.py
```
What will happen if an instance is deployed in a cloud?
During the first boot, it gets an IP address by DHCP before the full config is loaded. Then, after the address is obtained via DHCP, based on the metadata, it gets the other router's config.
Not sure what will be with `wait_for_commit_lock()` or `sudo /usr/libexec/vyos/conf_mode/protocols_static.py` in this case. Will this hook fail?

Good example for the working hook https://github.com/vyos/vyos-1x/blob/current/src/etc/dhcp/dhclient-enter-hooks.d/05-vyos-mtureplace

sever-sever commented:
@MattKobayashi The smoketest fails, could you re-check?
```
 DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py
DEBUG - test_01_static (__main__.TestProtocolsStatic.test_01_static) ... ok
DEBUG - test_02_static_table (__main__.TestProtocolsStatic.test_02_static_table) ... ok
DEBUG - test_03_static_vrf (__main__.TestProtocolsStatic.test_03_static_vrf) ... ok
DEBUG - test_04_static_multicast (__main__.TestProtocolsStatic.test_04_static_multicast) ... ok
DEBUG - test_05_dhcp_default_route (__main__.TestProtocolsStatic.test_05_dhcp_default_route) ... FAIL
DEBUG - test_06_dhcp_default_route_for_vrf (__main__.TestProtocolsStatic.test_06_dhcp_default_route_for_vrf) ... FAIL
DEBUG - test_06_dhcp_default_route_for_vrf (__main__.TestProtocolsStatic.test_06_dhcp_default_route_for_vrf) ... ERROR
DEBUG - test_07_dhcp_interface_static_routes (__main__.TestProtocolsStatic.test_07_dhcp_interface_static_routes) ... ERROR
DEBUG - test_07_dhcp_interface_static_routes (__main__.TestProtocolsStatic.test_07_dhcp_interface_static_routes) ... ERROR
DEBUG - 
DEBUG - ======================================================================
DEBUG - ERROR: test_06_dhcp_default_route_for_vrf (__main__.TestProtocolsStatic.test_06_dhcp_default_route_for_vrf)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py", line 186, in tearDown
DEBUG -     self.cli_commit()
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/base_vyostest_shim.py", line 103, in cli_commit
DEBUG -     out = self._session.commit()
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 305, in commit
DEBUG -     out = self.__run_command([COMMIT])
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 235, in __run_command
DEBUG -     raise ConfigSessionError(output)
DEBUG - vyos.configsession.ConfigSessionError: [ vrf ]
DEBUG - Can not remove VRF "red", it still has member interfaces!
DEBUG - delete [ vrf ] failed
DEBUG - Commit failed
DEBUG - 
DEBUG - 
DEBUG - ======================================================================
DEBUG - ERROR: test_07_dhcp_interface_static_routes (__main__.TestProtocolsStatic.test_07_dhcp_interface_static_routes)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py", line 669, in test_07_dhcp_interface_static_routes
DEBUG -     self.cli_commit()
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/base_vyostest_shim.py", line 103, in cli_commit
DEBUG -     out = self._session.commit()
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 305, in commit
DEBUG -     out = self.__run_command([COMMIT])
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 235, in __run_command
DEBUG -     raise ConfigSessionError(output)
DEBUG - vyos.configsession.ConfigSessionError: [ vrf ]
DEBUG - Can not remove VRF "red", it still has member interfaces!
DEBUG - delete [ vrf ] failed
DEBUG - Commit failed
DEBUG - 
DEBUG - 
DEBUG - ======================================================================
DEBUG - ERROR: test_07_dhcp_interface_static_routes (__main__.TestProtocolsStatic.test_07_dhcp_interface_static_routes)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py", line 186, in tearDown
DEBUG -     self.cli_commit()
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/base_vyostest_shim.py", line 103, in cli_commit
DEBUG -     out = self._session.commit()
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 305, in commit
DEBUG -     out = self.__run_command([COMMIT])
DEBUG -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DEBUG -   File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 235, in __run_command
DEBUG -     raise ConfigSessionError(output)
DEBUG - vyos.configsession.ConfigSessionError: [ vrf ]
DEBUG - Can not remove VRF "red", it still has member interfaces!
DEBUG - delete [ vrf ] failed
DEBUG - Commit failed
DEBUG - 
DEBUG - 
DEBUG - ======================================================================
DEBUG - FAIL: test_05_dhcp_default_route (__main__.TestProtocolsStatic.test_05_dhcp_default_route)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py", line 601, in test_05_dhcp_default_route
DEBUG -     self.assertIn(rf'ip route 0.0.0.0/0 {router} {interface} tag 210 {default_distance}', frrconfig)
DEBUG - AssertionError: 'ip route 0.0.0.0/0 192.0.2.2 eth0 tag 210 210' not found in 'frr version 10.2.4\nfrr defaults traditional\nhostname vyos\nlog syslog\nlog facility local7\nservice integrated-vtysh-config\n!\nip prefix-list pfx-foo-in seq 10 permit 192.0.2.0/25\nip prefix-list pfx-foo-out seq 10 permit 192.0.2.128/25\n!\nipv6 prefix-list pfx-foo-in6 seq 10 permit 2001:db8:1000::/64\nipv6 prefix-list pfx-foo-out6 seq 10 deny 2001:db8:2000::/64\n!\nvrf VRF-1\nexit-vrf\n!\nvrf VRF-2\nexit-vrf\n!\nvrf LPBRVRF\nexit-vrf\n!\nvrf PBRVRF\nexit-vrf\n!\nvrf red\nexit-vrf\n!\nvrf blue\nexit-vrf\n!\nvrf green\nexit-vrf\n!\nvrf black\nexit-vrf\n!\nroute-map foo-map-in permit 10\nexit\n!\nroute-map foo-map-out permit 10\nexit\n!\nroute-map redistr-ipv4-table-10 permit 10\nexit\n!\nroute-map redistr-ipv4-table-20 permit 10\nexit\n!\nroute-map redistr-ipv4-table-30 permit 10\nexit\n!\nroute-map redistr-ipv6-table-110 permit 10\nexit\n!\nroute-map redistr-ipv6-table-120 permit 10\nexit\n!\nroute-map redistr-ipv6-table-130 permit 10\nexit\n!\nend'
DEBUG - 
DEBUG - ======================================================================
DEBUG - FAIL: test_06_dhcp_default_route_for_vrf (__main__.TestProtocolsStatic.test_06_dhcp_default_route_for_vrf)
DEBUG - ----------------------------------------------------------------------
DEBUG - Traceback (most recent call last):
DEBUG -   File "/usr/libexec/vyos/tests/smoke/cli/test_protocols_static.py", line 644, in test_06_dhcp_default_route_for_vrf
DEBUG -     self.assertIn(rf'ip route 0.0.0.0/0 {router} {interface} tag 210 {default_distance}', frrconfig)
DEBUG - AssertionError: 'ip route 0.0.0.0/0 192.0.2.2 eth0 tag 210 210' not found in 'vrf red\nexit-vrf'
DEBUG - 
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 7 tests in 105.667s
DEBUG - 
DEBUG - FAILED (failures=2, errors=3)
DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_protocols_static_arp.py
DEBUG - test_static_arp (__main__.TestARP.test_static_arp) ... ok
DEBUG - 
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 1 test in 6.730s
DEBUG - 
```

sarthurdev reviewed (approved):
Logic looks sound, based on existing ipsec hook and includes smoketest. Haven't tested in build.

sever-sever commented:
@MattKobayashi could you re-check if it passed smoketests?

sever-sever commented:
@MattKobayashi, any updates?

MattKobayashi committed:
T3680: protocols: add dhclient hooks for dhcp-interface static routes

MattKobayashi commented:
@sever-sever My apologies, I've been busy with other projects. I believe I've resolved the smoketest issue, one is running now.

sever-sever commented:
@hedrok could you check and approve this PR if all works fine?
Until we find/implement/check another solution, I'm happy with this one.
Plus, it will be easier to backport.

sever-sever reviewed (approved):
We already have a similar logic for [ipsec-dhclient-hook](https://github.com/vyos/vyos-1x/blob/current/src/etc/dhcp/dhclient-exit-hooks.d/99-ipsec-dhclient-hook)
I don't expect any harm here.

hedrok commented:
Thanks for your work and patience.
I've tested these changes and all three issues in https://vyos.dev/T3680 are still present.

I've tried simple topology
    vyos-server [VyOS]
    vyos-client [VyOS]
    vyos-server.eth0 -- vyos-client.eth0

With configurations:

vyos-server:
```
    configure
    set system host-name vyos-server

    set interfaces ethernet eth0 address '10.1.1.1/24'
    set interfaces ethernet eth0 description 'LAN'

    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 option default-router '10.1.1.1'
    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 option name-server '10.1.1.1'
    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 option domain-name 'vyos'
    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 lease '86400'
    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 range 0 start '10.1.1.9'
    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 range 0 stop '10.1.1.254'
    set service dhcp-server shared-network-name LAN subnet 10.1.1.0/24 subnet-id '1'
    
    set service dns forwarding cache-size '0'
    set service dns forwarding listen-address '10.1.1.1'
    set service dns forwarding allow-from '10.1.1.0/24'

    set interfaces ethernet eth1 address '10.1.2.1/24'
    commit comment "vyos-server"
    save
```

vyos-client
```
    configure
    set system host-name vyos-client
    set interfaces ethernet eth0 address dhcp
    set protocols static route 10.1.2.0/24 dhcp-interface 'eth0'
    commit
    save
```

With such configuration I can still reproduce all 3 issues, to reproduce 2nd one I've deleted vyos-server and created new one with 10.147.1 instead of 10.1.1, but client after that still has
```
ip route 10.1.2.0/24 10.1.1.1 eth0
ip route 10.1.4.0/24 10.1.1.1 eth0
```

In `/etc/frr/frr.conf`.

I think the reason is that FRRender (`python/build/lib/vyos/frrender.py`) caches previous config and regenerates FRR configuration only if something has changed in VyOS configuration, so when you call `/usr/libexec/vyos/conf_mode/protocols_static.py` in dhclient hook FRR configuration is not regenerated. (I haven't checked this theory).

Do you check your changes on VyOS-rolling release?

Do I understand correctly that this PR should fix the issue

> 2. Renewing the DHCP lease after changing the default-router option or renumbering the network doesn't update the dhcp-interface next-hop in FRR.

?

If not, please describe what exactly this PR fixes so that I can check that, thanks again.

MattKobayashi commented:
Hi @hedrok,

>With such configuration I can still reproduce all 3 issues, to reproduce 2nd one I've deleted vyos-server and created new one with 10.147.1 instead of 10.1.1

Did you renew DHCP on the client interface after changing the server subnet? DHCP won't instantly renew under such conditions, you have to trigger it with `renew dhcp interface eth0` for the change to be picked up (or wait for expiry / 2 per RFC 2131 4.4.5).

>I think the reason is that FRRender (python/build/lib/vyos/frrender.py) caches previous config and regenerates FRR configuration only if something has changed in VyOS configuration, so when you call /usr/libexec/vyos/conf_mode/protocols_static.py in dhclient hook FRR configuration is not regenerated. (I haven't checked this theory).

I can't identify any logic in `protocols_static.py` that caches the previous FRR configuration, so I don't believe this theory holds.

>Do I understand correctly that this PR should fix the issue

>2. Renewing the DHCP lease after changing the default-router option or renumbering the network doesn't update the dhcp-interface next-hop in FRR.

Your understanding is correct. For total clarity:

>1. Setting the interface to DHCP and adding a dhcp-interface static route in the same commit doesn't add the route to FRR.

This can't be easily solved as the DHCP hook gets called before the DHCP interface list at `/tmp/static_dhcp_interfaces` is created. A second DHCP renewal after the commit should catch this, but I don't know of an easy way to catch it during the commit.

>3. Rebooting after changing the default-router option or renumbering the network leaves the old dhcp-interface next-hop in FRR.

This issue will also occur as a result of the DHCP hook being called before the interface list file is generated (same problem as 1).

hedrok committed:
T5811: Make static dhcp-interface routes robust

Solves the problem that vyos-configs in FRRender caches configuation and
DHCP changes are ignored.

* Add src/helpers/vyos-request-configd-update.py that requests vyos-configd
  to update FRR configuration.
* Make dhclient hooks use it instead of calling protocols_static.py
* Make FRRender cache not only configuration but also DHCP gateways so
  that is any of them changes, FRR configuration is updated

MattKobayashi commented:
@hedrok I've done some work on a new approach that I believe may solve the entire original issue. Please take a look and give me your thoughts: https://github.com/MattKobayashi/vyos-1x/compare/current...MattKobayashi:vyos-1x:T3680-dev

hedrok commented:
I've almost finalized an update to your previous solution: https://github.com/vyos/vyos-1x/commit/856a88e1b61d4eb70001db941421a27948b85663

At a glance my variant updates vyos-configd/FRRender so that it takes into account that dhcp interfaces could have changed and yours forces protocols_static to apply FRR configuration without `vyos-configd`.

I like my option more as if other places (not only protocols_static) require updates according to DHCP update it will be trivial to update my solution so that everything is done through `vyos-configd` in one go.
Plus it is just harder to predict what happens when most of FRR changes go through vyos-configd and in this specific case are applied directly in protocols_static.

But anyway my plan is:
1) Finalize my solution. (after this I'll write comment here so that you can take a look and test my variant if you have time).
2) Review your update more thougroughly.
3) Write here comparison, pros and cons of both options so that other reviewers decide which is better.

hedrok commented:
I've finalized and tested my solution, it has two commits:
* https://github.com/MattKobayashi/vyos-1x/commit/6db059860facd09a2a4bc80de184eebbfad28ef5 - uses `.lease` file instead of `.leases` file as I've seen it take wrong lease from that one.
* https://github.com/MattKobayashi/vyos-1x/commit/75e31efbd632562365f010d2a254cc7257790bf7 - main changes to your PR. There is detailed description.

Fixes all three subissues.

Now I'll review your new changes, I'll write back on Monday latest. If you have a chance to take a look and test my commits, it would be great.

For review convienience I've created PR to your branch in your repo: https://github.com/MattKobayashi/vyos-1x/pull/1
But please write overall comments here so that there is one place with whole discussion.

hedrok commented:
High-level review of suggested changes:

1. This is important update, we need not to loose this even if my solution is taken:
```
-    if [ "$old_ip_address" == "$new_ip_address" ]; then
+    if [ "$old_ip_address" == "$new_ip_address" ] && [ "$old_routers" == "$new_routers" ]; then
```
We can even leave just
```
-    if [ "$old_ip_address" == "$new_ip_address" ]; then
+    if [ "$old_routers" == "$new_routers" ]; then
```
As IP address of the interface is not important for update.

2. Test: for more realistic testing you can use virtual-ethernet to lauch dhcp server and client on same machine. You can see example here https://github.com/vyos/vyos-1x/pull/4783/files#diff-0ae71b5cc4b1cf4fa2b5112c8c75309271752e98769b349ce3d3d9a252dbd08a (still under review). Personally I am ok with mocking, but other maintainers prefer more realistic tests and in this particular case I've updated function that gets router IP so that it uses another file.

3. File with interfaces vs reading config

Personally I prefer previous version: reading whole config is relatively costly operation + in such case when adding `dhcp` somewhere in configuration we'll have to remember to update hooks, which can lead to bugs in the future.

4. Force apply: as I've written above, it seems risky to me. I would prefer using `vyos-configd`.

So if you are ok with my version, you can merge my PR, add your change in 1 with `"$old_routers" == "$new_routers"` as separate commit and I hope we'll be able to merge this PR soon. If not - please write details.

Thanks for your time.

MattKobayashi commented:
Hi @hedrok, your changes seem like a more robust solution to me, so I've merged your PR and added a commit to keep the comparison of routers instead of IP addresses. Hopefully this can be merged soon!

hedrok commented:
Only test
```
2025-10-21T00:52:52.9344085Z DEBUG - test_pppoe_client (__main__.PPPoEInterfaceTest.test_pppoe_client) ... FAIL
```
fails which is known issue in FRR.

@MattKobayashi thanks for fixing my linting issues - sorry you had to do that.

hedrok commented:
I've rebased branch, squashed most commits.

Rechecked everything, fixes all issues in https://vyos.dev/T5811 and https://vyos.dev/T3680:

1. Simultanious dhcp address + dhcp route: OK
2. After reboot, DHCP server down, then up: OK
3. DHCP gateway network change (prefix changes): OK
4. DHCP gateway only change (IP and network stay same): OK

CI for test_interfaces_cli failed only for known failure:

```
2025-10-22T09:34:00.8294552Z DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_interfaces_pppoe.py
...
2025-10-22T09:34:20.1811897Z DEBUG - test_pppoe_client (__main__.PPPoEInterfaceTest.test_pppoe_client) ... FAIL
```

@sever-sever please re-review and merge.

hedrok commented:
As I'm not an author of PR, I can't resolve comments from CoPilot, so I'll reply here.
All issues can be ignored in my opinion.
```
print(f"Received reply {request} [ {message} ]")
```
Yes, maybe it is extra information, but it is just copied from official `zmq` documentation and I see no harm in it: https://zeromq.org/get-started/?language=python#
Plus in such way you see in one line request and message.

As for using `=` or `[[` in bash - I didn't check, maybe it would be better, but we should change it in all scripts then, as in most of VyOS dhclient hooks comparisons are same as here.

hedrok commented:
@MattKobayashi thanks again, merged at last :)
